Add support for array logging and json format

MaticPogacnik opened the pull request:
Features in this PR are based on the: https://gist.github.com/anjohnson/eee1d8fce75bb406cf2245e2d8cc274e

New backend was written to support logging in a JSON format (requires EPICS base 7.0.1 or newer). 
It also supports arrays and lso/lsi records now.

This is work in progress and is not ready to merge.  But the code will not change much from this point.
Any feedback would be greatly appreciated.

TODO:

- [x] Document code (doxygen)
- [x] Add user documentation
- [x] Add unit tests.

anjohnson commented on caPutLogApp/caPutJsonLogTask.cpp:
```diff
@@ -0,0 +1,711 @@
+    yajl_gen_get_buf(handle, &buf, &len);
+
+    /* Get a JSON as a string */
+    std::string json (reinterpret_cast<const char *>(buf));
```
Is it intentional that the `logToServer()` method modifies that string (appending a newline), so the `logToPV()` method gets the string with the newline added?

MaticPogacnik replied:
Yes, this was a bug. Fixed in: f4683afa4228782f2603c05d58bce7f15055a0cc
The whole purpose why I didn't append new line before was to have log inside a PV without a new line.

MaticPogacnik commented on caPutLogApp/caPutLogTask.h:
```diff
@@ -12,6 +13,14 @@ extern "C" {
 #define MAX_USERID_SIZE 32
 #define MAX_HOSTID_SIZE 32
 
+#define JSON_AND_ARRAYS_SUPPORTED (EPICS_VERSION * 10000 + EPICS_REVISION * 100 + EPICS_MODIFICATION >= 70001)
```
This make sense yes, I moved the logic to the Makefile in: 9a8aee1

MaticPogacnik commented on caPutLogApp/caPutLogAs.c:
```diff
@@ -170,10 +198,47 @@ static void caPutLogAs(asTrapWriteMessage *pmessage, int afterPut)
 }
 
+int caPutLogMaxArraySize(short type)
+{
```
I changed this in: deac70395ab98d312eaea0a296523447bb71674c
Function is still required as I do some small checks. But array is static in the method so its should be ok.

MaticPogacnik commented on caPutLogApp/caPutJsonLogTask.h:
```diff
@@ -0,0 +1,274 @@
+
+// Standard library imports
+#include <string>
+#include <atomic>
```
Changed in: d8659d537f587ac7da17dc5905dde3a9a7e90d8e
Yes, I believe that maintaining this many platform can be hard.

Is there any general rule with which standard code should be compatible with?

anjohnson replied:
Thanks, looks good.

EPICS 7.1.x versions will require a C++11 compiler, but we don't have a specific plan for when we're going to get there, it generally depends on the community's needs and who's willing to work on the code. We do run our CI tests on a C++11 build now (thanks to Travis-CI) but some of our targets aren't there yet.

BTW I tried building this branch on my laptop, and the Clang compiler's default C++ standard doesn't like your 'noexcept' markings.

MaticPogacnik replied:
I removed `noexcept`, move constructor and `nullptr`, for better compatability, as these are all c++11 features.

MaticPogacnik commented on caPutLogApp/caPutJsonLogTask.h:
```diff
@@ -0,0 +1,274 @@
+ * For the instructions how to use the utility please refer to the user manual.
+ *
+ */
+class CaPutJsonLogTask {
```
Added in: 47a5a12593d0b36a99258ec7953232156551b956
Sorry about this, I completely forgot about this, on the Linux where I developed it worked. :)
How do you test base code on all the supported platforms? CI ?

anjohnson replied:
You can give gcc some command-line flags which make it only export symbols from a shared library which are explicitly marked for export. The shareLib.h header provides the necessary symbol decorations for gcc on Linux as well as for Windows, and this is very useful for finding symbol decoration issues. Base doesn't give that by default, but I have a file ~/configure/CONFIG_USER.Common.linux-x86_64 which contains:
```Makefile
OP_SYS_CFLAGS += -Werror=pointer-arith
OP_SYS_CFLAGS += -fvisibility=hidden
OP_SYS_CXXFLAGS += -fvisibility-inlines-hidden
```
Yes, the EPICS build system does look in your $HOME/configure directory for specific files which you can use to customize builds, so the above flags appear on my compiler command-lines whenever I build for `linux-x86_64`.

MaticPogacnik commented on caPutLogApp/caPutLogAs.c:
```diff
@@ -140,9 +151,17 @@ static void caPutLogAs(asTrapWriteMessage *pmessage, int afterPut)
+            plogData->old_size = noElements;
+        } else {
+            plogData->old_size = num_elm;
+        }
```
I changed the logic how is determined if a value is an array here: b5a8e866b446674a4bdde15b328752ac3a5a10c9

I extracted it to a separate function instead of putting it before the `if` statement, as that would require more changes to existing code (most pointers with required information are obtained inside the `if` statement ).

MaticPogacnik commented on caPutLogApp/caPutJsonLogTask.cpp:
```diff
@@ -0,0 +1,711 @@
+}
+
+CaPutJsonLogTask::CaPutJsonLogTask()
+    : caPutJsonLogQ(1000, sizeof(LOGDATA *)),
```
Old implementation has this hard coded here: https://github.com/epics-modules/caPutLog/blob/57dc99a8a89381e3d5b4a4b02e9f8b5645b2a8be/caPutLogApp/caPutLogAs.c#L68

But yes, this could be useful. Added in: cc1eb3a8faf49c39441a0aa2670eb452757a298b

MaticPogacnik replied:
This variable must be set before the `caPutJsonLogInit` call. I will add this to the documentation. Default value is still 1000.

MaticPogacnik commented on caPutLogApp/caPutJsonLogTask.cpp:
```diff
@@ -0,0 +1,711 @@
+    if (pLogData->new_size == 0) {
+        const unsigned char empty_value[] = "<empty>";
+        CALL_YAJL_FUNCTION_AND_CHECK_STATUS(status, yajl_gen_string(handle, empty_value,
+                            strlen(reinterpret_cast<const char *>(empty_value))));
```
Changed in: deac70395ab98d312eaea0a296523447bb71674c

MaticPogacnik commented on caPutLogApp/caPutJsonLogTask.cpp:
```diff
@@ -0,0 +1,711 @@
+    // Arrays and scalars (all except DBF_CHAR)
+    else {
+        for (int i = 0; i < pLogData->new_log_size; i++) {
+            fieldVal2Str(reinterpret_cast<char *>(interBuffer), interBufferSize,
```
Yes! I missed this.
Changed here: 901b4beb68f87a9ac3747c07c68bccd01b82f6c5

I was not able to use `isinf` as it does not distinguish between minus and plus infinity. But comparing to `epicsINF` and `-epicsINF` works.

MaticPogacnik commented on caPutLogApp/caPutJsonLogTask.cpp:
```diff
@@ -0,0 +1,711 @@
+    }
+
+    // Register exit handler
+    epicsAtExit(caPutJsonLogExit, NULL);
```
Yes, the pointer (and the object of course) is there for the lifetime of the IOC process and can be started again if required, but this is not implemented (can be added if needed). For example, I can add two iocsh commands like `caPutJsonLogStart` and `caPutJsonLogStop`, which can be used to start/stop the logger if required.

Currently I just call stop on exit, so the stopper flag is set, and the logger thread exits.

MaticPogacnik commented:
Andrew,

Thank you for the feedback, I resolved the comments and added links to commits for easier navigation.

I still have a problem with logging to a PV, if I use a waveform records it works, but with lso record i have some problems.

This doesn't seem to work: 
```
status = dbPutField(this->pCaPutJsonLogPV, DBR_CHAR, msg.c_str(), msg.length());
```
Should `DBR_CHAR` be something else?

anjohnson commented:
This is how to log to both waveform and long string PVs:
```diff
-        status = dbPutField(this->pCaPutJsonLogPV, DBR_CHAR, msg.c_str(), msg.length());
+        if (this->pCaPutJsonLogPV->field_type == DBR_CHAR)
+            status = dbPutField(this->pCaPutJsonLogPV, DBR_CHAR, msg.c_str(), msg.length());
+        else
+            status = dbPutField(this->pCaPutJsonLogPV, DBR_STRING, msg.c_str(), 1);
```
Even though it's using `DBR_STRING` the actual string can be longer than 40 characters.

anjohnson commented:
This is how to stop the RSET build warnings:
```diff
diff --git a/caPutLogApp/Makefile b/caPutLogApp/Makefile
index 3b9f718..b9272f9 100644
--- a/caPutLogApp/Makefile
+++ b/caPutLogApp/Makefile
@@ -5,6 +5,8 @@ include $(TOP)/configure/CONFIG
 
 LIBRARY_IOC = caPutLog
 
+USR_CPPFLAGS += -DUSE_TYPED_RSET
+
 caPutLog_SRCS += caPutLogTask.c
 caPutLog_SRCS += caPutLogAs.c
 caPutLog_SRCS += caPutLogClient.c
```

MaticPogacnik commented:
> This is how to log to both waveform and long string PVs:
> 
> ```diff
> -        status = dbPutField(this->pCaPutJsonLogPV, DBR_CHAR, msg.c_str(), msg.length());
> +        if (this->pCaPutJsonLogPV->field_type == DBR_CHAR)
> +            status = dbPutField(this->pCaPutJsonLogPV, DBR_CHAR, msg.c_str(), msg.length());
> +        else
> +            status = dbPutField(this->pCaPutJsonLogPV, DBR_STRING, msg.c_str(), 1);
> ```
> 
> Even though it's using `DBR_STRING` the actual string can be longer than 40 characters.

I tried this but I am getting only first 39 (+ terminator) writen to lso record (`\` escape characters must be removed to get this number):

![image](https://user-images.githubusercontent.com/5747624/90023538-2323e380-dcb4-11ea-8f33-3faf9ede10f4.png)
I checked with the debugger and this line is exectued: https://github.com/epics-base/epics-base/blob/951b6acbbc102a5865592f671d2b410e96c5a4d3/modules/database/src/ioc/db/dbFastLinkConv.c#L85 which is also the reason why the string is clipped?

In the units tests I had to use `DBR_CHAR` to make it work (tho one is over CA and one not):

```
    chid pchid;
    const char *lsoPv = "lso.$";
    SEVCHK(ca_create_channel(lsoPv, NULL, NULL, 0, &pchid), "ca_create_channel failed");
    SEVCHK(ca_pend_io (CA_PEND_IO_TIMEOUT), "ca_pend_io failed");
    std::string putValue1("lsostring1234567890123456789012345678901234567890123456789");
    SEVCHK(ca_array_put(DBR_CHAR, putValue1.length(), pchid, (void *) putValue1.c_str()), "ca_array_put error");
        SEVCHK(ca_pend_io (CA_PEND_IO_TIMEOUT), "ca_pend_io error");
```
I took `caput` client as an example: https://github.com/epics-base/epics-base/blob/951b6acbbc102a5865592f671d2b410e96c5a4d3/modules/ca/src/tools/caput.c#L507
I believe this code is executed if `-S` flag is specified.

anjohnson commented:
I suspect your `caget loglso` is what is limiting the length of the string you are seeing. Try using `caget -S loglso.$` instead, does that clip the string too?

MaticPogacnik commented:
I tried it and this is the output:
```
[devman@anl-develop ~]$ caget -S loglso.$
loglso.$ {\"date\":\"12-08-2020\",\"time\":\"16:57:08\",
[devman@anl-develop ~]$ caget loglso.LEN
loglso.LEN                     40
```
I also checked the `LEN` field and it confirm that it's only 40 characters long.

anjohnson commented:
Hi Matic,

I get these build warnings from clang (on macOS), which should be easily fixed:
```
woz$ make -sj
../caPutJsonLogTest.cpp:189:18: warning: in-class initialization of non-static
      data member is a C++11 extension [-Wc++11-extensions]
    bool inArray = false;
                 ^
../caPutJsonLogTest.cpp:190:20: warning: in-class initialization of non-static
      data member is a C++11 extension [-Wc++11-extensions]
    int arrayCount = 0;
                   ^
../caPutJsonLogTest.cpp:191:18: warning: in-class initialization of non-static
      data member is a C++11 extension [-Wc++11-extensions]
    bool isArray = false;
                 ^
../caPutJsonLogTest.cpp:193:21: warning: in-class initialization of non-static
      data member is a C++11 extension [-Wc++11-extensions]
    bool waitingKey = true;
                    ^
../caPutJsonLogTest.cpp:210:17: warning: in-class initialization of non-static
      data member is a C++11 extension [-Wc++11-extensions]
    int newSize = -1;
                ^
../caPutJsonLogTest.cpp:211:17: warning: in-class initialization of non-static
      data member is a C++11 extension [-Wc++11-extensions]
    int oldSize = -1;
                ^
../caPutJsonLogTest.cpp:190:9: warning: private field 'arrayCount' is not used
      [-Wunused-private-field]
    int arrayCount = 0;
        ^
../caPutJsonLogTest.cpp:191:10: warning: private field 'isArray' is not used
      [-Wunused-private-field]
    bool isArray = false;
         ^
8 warnings generated.
```

To build successfully against Base-3.15 I had to add
```C
#include <recSup.h>
```
to the file caPutLogApp/caPutLogAs.c

caPutLogApp/caPutLogTask.h doesn't need to `#include <epicsVersion.h>` any more since you're using the Makefile-defined macro instead.

Thank-you for fixing the documentation, that looks much better when viewed on GitHub now. I haven't tried building the docs using sphinx as I don't have it installed, have you attempted that at all?

In test/Makefile the PROD_LIBS need to be listed in the reverse order for the test code to link properly when building a static binary. The library order doesn't matter when linking against shared libraries or DLLs, but for static linking `Com` should always come last or the link will fail. Actually you don't have to list all of the Base libraries anyway, the build system defines a variable with all the IOC libraries in it, so do this instead:
```makefile
PROD_LIBS = caPutLog
PROD_LIBS += $(EPICS_BASE_IOC_LIBS)
```

The tests generally look good, but there do seem to be a lot of 10 second delays in between bursts of checks, are the delays included to avoid updates generating bursts? If they are required that's fine, I don't need the test code to run any faster, I'm just wondering what's going on while the program is paused.

With the above final fixes I think this looks good and mergeable, thanks!

I will see you at the meeting with Guobao on Wednesday.

MaticPogacnik commented:
Thank you for review Andrew,

> I get these build warnings from clang (on macOS), which should be easily fixed:
> 
> ```
> woz$ make -sj
> ../caPutJsonLogTest.cpp:189:18: warning: in-class initialization of non-static
>       data member is a C++11 extension [-Wc++11-extensions]
>     bool inArray = false;
>                  ^
> ../caPutJsonLogTest.cpp:190:20: warning: in-class initialization of non-static
>       data member is a C++11 extension [-Wc++11-extensions]
>     int arrayCount = 0;
>                    ^
> ../caPutJsonLogTest.cpp:191:18: warning: in-class initialization of non-static
>       data member is a C++11 extension [-Wc++11-extensions]
>     bool isArray = false;
>                  ^
> ../caPutJsonLogTest.cpp:193:21: warning: in-class initialization of non-static
>       data member is a C++11 extension [-Wc++11-extensions]
>     bool waitingKey = true;
>                     ^
> ../caPutJsonLogTest.cpp:210:17: warning: in-class initialization of non-static
>       data member is a C++11 extension [-Wc++11-extensions]
>     int newSize = -1;
>                 ^
> ../caPutJsonLogTest.cpp:211:17: warning: in-class initialization of non-static
>       data member is a C++11 extension [-Wc++11-extensions]
>     int oldSize = -1;
>                 ^
> ../caPutJsonLogTest.cpp:190:9: warning: private field 'arrayCount' is not used
>       [-Wunused-private-field]
>     int arrayCount = 0;
>         ^
> ../caPutJsonLogTest.cpp:191:10: warning: private field 'isArray' is not used
>       [-Wunused-private-field]
>     bool isArray = false;
>          ^
> 8 warnings generated.
> ```

I fixed this and removed some other `c++11` features as well in c9153ac. Code compiles without any warnings with `-std=c++98` flag now.

> The tests generally look good, but there do seem to be a lot of 10 second delays in between bursts of checks, are the delays included to avoid updates generating bursts? If they are required that's fine, I don't need the test code to run any faster, I'm just wondering what's going on while the program is paused.

Yes, there is a 5 seconds timeout in the logic, to see if there is any more puts coming. Test code makes 1 caput, and then waits for the log to arrive, but the logger will not send one until that 5 seconds timer expires. I added some `testDiag` to the tests, with this message.

> Thank-you for fixing the documentation, that looks much better when viewed on GitHub now. I haven't tried building the docs using sphinx as I don't have it installed, have you attempted that at all?

Yes, I did build the web version (I do not have latex installed for pdf, ...). Here is compressed build `docs` directory: 
[docs.tar.gz](https://github.com/epics-modules/caPutLog/files/5088940/docs.tar.gz)

Best regards, 
Matic
